README.md
# pararius-apartments-alert
Scraping and notification application for pararius.nl

##
NOTICE: this is a personal project and it has not been/is not being developed with a pleasasant nor straightforward user experience in mind. The code is also not following industry standards because there is no need to. I'm developing this for fun to see if I can reproduce the service [Pararius+](https://www.pararius.nl/info/parariusplus) provides. It is not meant to substitute it nor compete with it.

## Requirements
Python libraries: 'bs4', 'selenium', 'webdriver-manager','gtts', 'telegram_send', 'pysimplegui'

Manual install with 
`pip install LIBRARY_NAME`
or run 
`python3 ./libraries/install_dependencies.py`

## RUN
`python3 ./main.py`

## Known issues and future updates
- Installation of the required python packages is manual or via the script mentione above. Future updates will detect missing libraries and install them automatically.
- No cancel button implementation yet. To parse another url simply paste the new one an the thread in the background will pick it up, but the only way to not parse is to close the app





debian/changelog
pararius-apartments-alert (0.1) unstable; urgency=medium

  * Initial release. (Closes: #XXXXXX)

 -- Luca Lagunas <lagunasluca@protonmail.com>  Sun, 12 Nov 2023 14:28:16 +0100


test.py
import PySimpleGUI as sg

def make_window(theme):
    sg.theme(theme)
    menu_def = [['&Application', ['E&xit']],
                ['&Help', ['&About']] ]
    right_click_menu_def = [[], ['Edit Me', 'Versions', 'Nothing','More Nothing','Exit']]
    graph_right_click_menu_def = [[], ['Erase','Draw Line', 'Draw',['Circle', 'Rectangle', 'Image'], 'Exit']]

    # Table Data
    data = [["John", 10], ["Jen", 5]]
    headings = ["Name", "Score"]

    input_layout =  [

                # [sg.Menu(menu_def, key='-MENU-')],
                [sg.Text('Anything that requires user-input is in this tab!')], 
                [sg.Input(key='-INPUT-')],
                [sg.Slider(orientation='h', key='-SKIDER-'),
                 sg.Image(data=sg.DEFAULT_BASE64_LOADING_GIF, enable_events=True, key='-GIF-IMAGE-'),],
                [sg.Checkbox('Checkbox', default=True, k='-CB-')],
                [sg.Radio('Radio1', "RadioDemo", default=True, size=(10,1), k='-R1-'), sg.Radio('Radio2', "RadioDemo", default=True, size=(10,1), k='-R2-')],
                [sg.Combo(values=('Combo 1', 'Combo 2', 'Combo 3'), default_value='Combo 1', readonly=False, k='-COMBO-'),
                 sg.OptionMenu(values=('Option 1', 'Option 2', 'Option 3'),  k='-OPTION MENU-'),],
                [sg.Spin([i for i in range(1,11)], initial_value=10, k='-SPIN-'), sg.Text('Spin')],
                [sg.Multiline('Demo of a Multi-Line Text Element!\nLine 2\nLine 3\nLine 4\nLine 5\nLine 6\nLine 7\nYou get the point.', size=(45,5), expand_x=True, expand_y=True, k='-MLINE-')],
                [sg.Button('Button'), sg.Button('Popup'), sg.Button(image_data=sg.DEFAULT_BASE64_ICON, key='-LOGO-')]]

    asthetic_layout = [[sg.T('Anything that you would use for asthetics is in this tab!')],
               [sg.Image(data=sg.DEFAULT_BASE64_ICON,  k='-IMAGE-')],
               [sg.ProgressBar(100, orientation='h', size=(20, 20), key='-PROGRESS BAR-'), sg.Button('Test Progress bar')]]

    logging_layout = [[sg.Text("Anything printed will display here!")],
                      [sg.Multiline(size=(60,15), font='Courier 8', expand_x=True, expand_y=True, write_only=True,
                                    reroute_stdout=True, reroute_stderr=True, echo_stdout_stderr=True, autoscroll=True, auto_refresh=True)]
                      # [sg.Output(size=(60,15), font='Courier 8', expand_x=True, expand_y=True)]
                      ]
    
    graphing_layout = [[sg.Text("Anything you would use to graph will display here!")],
                      [sg.Graph((200,200), (0,0),(200,200),background_color="black", key='-GRAPH-', enable_events=True,
                                right_click_menu=graph_right_click_menu_def)],
                      [sg.T('Click anywhere on graph to draw a circle')],
                      [sg.Table(values=data, headings=headings, max_col_width=25,
                                background_color='black',
                                auto_size_columns=True,
                                display_row_numbers=True,
                                justification='right',
                                num_rows=2,
                                alternating_row_color='black',
                                key='-TABLE-',
                                row_height=25)]]

    popup_layout = [[sg.Text("Popup Testing")],
                    [sg.Button("Open Folder")],
                    [sg.Button("Open File")]]
    
    theme_layout = [[sg.Text("See how elements look under different themes by choosing a different theme here!")],
                    [sg.Listbox(values = sg.theme_list(), 
                      size =(20, 12), 
                      key ='-THEME LISTBOX-',
                      enable_events = True)],
                      [sg.Button("Set Theme")]]
    
    layout = [ [sg.MenubarCustom(menu_def, key='-MENU-', font='Courier 15', tearoff=True)],
                [sg.Text('Demo Of (Almost) All Elements', size=(38, 1), justification='center', font=("Helvetica", 16), relief=sg.RELIEF_RIDGE, k='-TEXT HEADING-', enable_events=True)]]
    layout +=[[sg.TabGroup([[  sg.Tab('Input Elements', input_layout),
                               sg.Tab('Asthetic Elements', asthetic_layout),
                               sg.Tab('Graphing', graphing_layout),
                               sg.Tab('Popups', popup_layout),
                               sg.Tab('Theming', theme_layout),
                               sg.Tab('Output', logging_layout)]], key='-TAB GROUP-', expand_x=True, expand_y=True),

               ]]
    layout[-1].append(sg.Sizegrip())
    window = sg.Window('All Elements Demo', layout, right_click_menu=right_click_menu_def, right_click_menu_tearoff=True, grab_anywhere=True, resizable=True, margins=(0,0), use_custom_titlebar=True, finalize=True, keep_on_top=True)
    window.set_min_size(window.size)
    return window

def main():
    window = make_window(sg.theme())

    # This is an Event Loop 
    while True:
        event, values = window.read(timeout=100)
        # keep an animation running so show things are happening
        if event not in (sg.TIMEOUT_EVENT, sg.WIN_CLOSED):
            print('============ Event = ', event, ' ==============')
            print('-------- Values Dictionary (key=value) --------')
            for key in values:
                print(key, ' = ',values[key])
        if event in (None, 'Exit'):
            print("[LOG] Clicked Exit!")
            break

        window['-GIF-IMAGE-'].update_animation(sg.DEFAULT_BASE64_LOADING_GIF, time_between_frames=100)
        if event == 'About':
            print("[LOG] Clicked About!")
            sg.popup('PySimpleGUI Demo All Elements',
                     'Right click anywhere to see right click menu',
                     'Visit each of the tabs to see available elements',
                     'Output of event and values can be see in Output tab',
                     'The event and values dictionary is printed after every event', keep_on_top=True)
        elif event == 'Popup':
            print("[LOG] Clicked Popup Button!")
            sg.popup("You pressed a button!", keep_on_top=True)
            print("[LOG] Dismissing Popup!")
        elif event == 'Test Progress bar':
            print("[LOG] Clicked Test Progress Bar!")
            progress_bar = window['-PROGRESS BAR-']
            for i in range(100):
                print("[LOG] Updating progress bar by 1 step ("+str(i)+")")
                progress_bar.update(current_count=i + 1)
            print("[LOG] Progress bar complete!")
        elif event == "-GRAPH-":
            graph = window['-GRAPH-']       # type: sg.Graph
            graph.draw_circle(values['-GRAPH-'], fill_color='yellow', radius=20)
            print("[LOG] Circle drawn at: " + str(values['-GRAPH-']))
        elif event == "Open Folder":
            print("[LOG] Clicked Open Folder!")
            folder_or_file = sg.popup_get_folder('Choose your folder', keep_on_top=True)
            sg.popup("You chose: " + str(folder_or_file), keep_on_top=True)
            print("[LOG] User chose folder: " + str(folder_or_file))
        elif event == "Open File":
            print("[LOG] Clicked Open File!")
            folder_or_file = sg.popup_get_file('Choose your file', keep_on_top=True)
            sg.popup("You chose: " + str(folder_or_file), keep_on_top=True)
            print("[LOG] User chose file: " + str(folder_or_file))
        elif event == "Set Theme":
            print("[LOG] Clicked Set Theme!")
            theme_chosen = values['-THEME LISTBOX-'][0]
            print("[LOG] User Chose Theme: " + str(theme_chosen))
            window.close()
            window = make_window(theme_chosen)
        elif event == 'Edit Me':
            sg.execute_editor(__file__)
        elif event == 'Versions':
            sg.popup_scrolled(__file__, sg.get_versions(), keep_on_top=True, non_blocking=True)

    window.close()
    exit(0)

if __name__ == '__main__':
    sg.theme('black')
    sg.theme('dark red')
    sg.theme('dark green 7')
    # sg.theme('DefaultNoMoreNagging')
    main()



pararius_alert_test.py
from libraries.pararius_alert import *
from libraries.file_management import *

from bs4 import BeautifulSoup
import time
import asyncio

def TestWithRemovedElement():
    
    #Should not find any apartment
    EXPECTED_OUTPUT = 0

    page_source = loadFileContent("test/pageSourceFull.txt")
    soup = BeautifulSoup(page_source, 'html.parser')
    oldRes = soup.find_all("h2", {"class": "listing-search-item__title"})

    page_source = loadFileContent("test/pageSourceTopElemRemoved.txt")
    soup = BeautifulSoup(page_source, 'html.parser')
    currentRes = soup.find_all("h2", {"class": "listing-search-item__title"})

    find_new_apartments(oldRes, currentRes)

    if find_new_apartments(oldRes, currentRes):
        TEST_OUTPUT = 1
    else:
        TEST_OUTPUT = 0

    return TEST_OUTPUT == EXPECTED_OUTPUT


def TestWithNewElement():
    
    #Should find a new apartment
    EXPECTED_OUTPUT = 1

    page_source = loadFileContent("test/pageSourceFull.txt")
    soup = BeautifulSoup(page_source, 'html.parser')
    oldRes = soup.find_all("h2", {"class": "listing-search-item__title"})

    page_source = loadFileContent("test/pageSourceNewElem.txt")
    soup = BeautifulSoup(page_source, 'html.parser')
    currentRes = soup.find_all("h2", {"class": "listing-search-item__title"})

    find_new_apartments(oldRes, currentRes)

    if find_new_apartments(oldRes, currentRes):
        TEST_OUTPUT = 1
    else:
        TEST_OUTPUT = 0

    return TEST_OUTPUT == EXPECTED_OUTPUT

def parametrizedParsePage(p_pageSource1, p_pageSource2):
    loop = asyncio.new_event_loop() 
    asyncio.set_event_loop(loop)

    print("Getting the page source")
    page_source = p_pageSource1
    soup = BeautifulSoup(page_source, 'html.parser')
    res = soup.find_all("h2", {"class": "listing-search-item__title"})

    TEST_i = 0

    #While True loop starts here in the actual function
    current_res = res

    page_source = p_pageSource2
    soup = BeautifulSoup(page_source, 'html.parser')
    res = soup.find_all("h2", {"class": "listing-search-item__title"})

    if res:

        if TEST_i == 0:
            print("Looking for changes. Stand by and wait for a notification")
            TEST_i += 1
        
        new_apartments = find_new_apartments(res, current_res)
        if (not new_apartments):
            time.sleep(1)
        else:
            #send notification
            loop = asyncio.get_event_loop()
            loop.run_until_complete(sendTelegramNotification(new_apartments, "TEST message to the broker"))

            #for logging purposes
            print('Nieuwe aanbieding gedetecteerd')

            time.sleep(1)

def testMain():
    #TEST 1
    print("Test 1: ", TestWithRemovedElement())
    #TEST 2
    print("Test 2: ", TestWithNewElement())


    #
    parametrizedParsePage(loadFileContent("test/pageSourceFull.txt"), loadFileContent("test/pageSourceNewElem.txt"))
testMain()

libraries/pararius_alert.py
from bs4 import BeautifulSoup
import time
import telegram_send
from selenium import webdriver
from selenium.webdriver.firefox.options import Options
import warnings
import asyncio

from selenium.webdriver.common.by import By
from selenium.webdriver.support.ui import WebDriverWait
from selenium.webdriver.support import expected_conditions as EC


def GetPageSource(p_userUrl):
    options = Options()
    options.add_argument('--ignore-certificate-errors')
    options.add_argument('--incognito')
    options.add_argument('--headless')
    with warnings.catch_warnings():
        warnings.filterwarnings("ignore", category=DeprecationWarning)
        driver = webdriver.Firefox(options=options)
    
    try:
        driver.set_page_load_timeout(120)
        driver.get(p_userUrl)

        WebDriverWait(driver, 120).until(EC.presence_of_element_located((By.CLASS_NAME, 'listing-search-item__title')))
        WebDriverWait(driver, 120).until(EC.presence_of_element_located((By.XPATH, '//body[not(@class="loading")]')))

        pageSource = driver.page_source
    except Exception as e:
        print(f"Failed to retrieve page source. Please check your connection..Error during page navigation: {e}")

        loop = asyncio.get_event_loop()
        loop.run_until_complete(sendTelegramNotification([], "Failed to retrieve page source. Please check your connection."))
        
        pageSource = None

    finally:
        driver.quit()
    
    return pageSource


async def sendTelegramNotification(p_apartmentsList, p_extraMessage):
        for property in p_apartmentsList:
            #By using the *property syntax, it unpacks the elements of the property list, and they will be passed as separate arguments to the send function.
            await telegram_send.send(messages=[("New property: ", *property)])
        await telegram_send.send(messages=[(p_extraMessage)])

def find_new_apartments(old_vacancies, updated_vacancies):
    ret = []
    # Exclude the last element
    if updated_vacancies and updated_vacancies[-1] not in old_vacancies:
        updated_vacancies = updated_vacancies[:-1]

    # print("------------------------DEBUG--------------------------")
    # print("old = ", old_vacancies[0].find('a').get_text(strip=True))
    # print("new = ", updated_vacancies[0].find('a').get_text(strip=True))
    # print("-------------------------------------------------------")
    new_apartments = set(updated_vacancies) - set(old_vacancies)
    
    for property in new_apartments:
        # print("property = ", property)
        name = property.find('a').get_text(strip=True)
        href = property.find('a')['href']
        ret.append((name, "https://www.pararius.nl" + href))
    return ret

def ParsePage(p_userUrl, p_messageToTheBroker, p_sleepBetweenSearches):
    loop = asyncio.new_event_loop() 
    asyncio.set_event_loop(loop)

    print("Getting the page source")
    page_source = GetPageSource(p_userUrl)
    soup = BeautifulSoup(page_source, 'html.parser')
    old_res = soup.find_all("h2", {"class": "listing-search-item__title"})

    TEST_i = 0
    DEBUG_state_index = 0

    while True:
        

        page_source = GetPageSource(p_userUrl)
        soup = BeautifulSoup(page_source, 'html.parser')
        # res = soup.find_all("h2", {"class": "listing-search-item__title"})
        current_res = soup.find_all("h2", {"class": "listing-search-item__title"})

        if current_res:
            # print("Still running. Check #", DEBUG_state_index)
            # DEBUG_state_index += 1
            if TEST_i == 0:
                print("Looking for changes. Stand by and wait for a notification")
                TEST_i += 1
            
            new_apartments = find_new_apartments(old_res, current_res)
            if (not new_apartments):
                # loop = asyncio.get_event_loop()
                # loop.run_until_complete(sendTelegramNotification(["test1", "no new partments"], p_messageToTheBroker))
                old_res = current_res
                time.sleep(p_sleepBetweenSearches)
            else:
                #send notification
                loop = asyncio.get_event_loop()
                loop.run_until_complete(sendTelegramNotification(new_apartments, p_messageToTheBroker))

                #for logging purposes
                print('Nieuwe aanbieding gedetecteerd')
                print(p_userUrl)

                old_res = current_res
                time.sleep(p_sleepBetweenSearches)

main.py
from threading import Thread
from libraries.pararius_alert import ParsePage
from libraries.file_management import *
import PySimpleGUI as sg

def main():
    sg.theme('SystemDefaultForReal')

    url_input = sg.InputText(key='-URL-', size=(60, 1))
    msg_input = [sg.Multiline(size=(60, 15), key='-MSG-', autoscroll=True, auto_refresh=True)]
    delay_input = sg.Input(key='-DELAY-', size=(5,1))

    parseLayout = [
        [sg.Text('Enter URL:')],
        [url_input],
        [sg.Text('Seconds between searches (default = 60) ')],
        [delay_input],
        [sg.Button('Parse', key='-PARSE-')],
        [sg.Multiline(size=(60,15), font='Courier 8', expand_x=True, expand_y=True, write_only=True,
                                    reroute_stdout=True, reroute_stderr=True, echo_stdout_stderr=True, autoscroll=True, auto_refresh=True)],
    ]

    telegramLayout = [
        [sg.Text('To setup your phone to receive notifications form the app follow the next steps:')],
        [sg.Text('1. Go to https://telegram.me/BotFather and send the message "/newbot" to the chat')],
        [sg.Text('2. Open the terminal and type "telegram-send --configure"')],
        [sg.Text('If the system does not recognise the command add your Python\\Script to PATH')],
        [sg.Text('3. Copy and paste the token you received from BotFather in the terminal')],
        [sg.Text('4. Now you will receive a notification every time the parser detects a new accomodation')],
        [sg.Text('Please visit the project\'s github page or the About section in the app to learn more about the app and its limitations')]
    ]

    messageLayout = [
        [sg.Text('Insert your message to broker. It will be sent as part of the notification so that you can copy and paste it quickly.')],
        [sg.Text('Leave it blank to use the previously inserted text')],
        msg_input
    ]

    aboutLayout = [
        [sg.Text('Scraping and notification application for pararius.nl\n')],
        [sg.Text('Known issues and future updates:')],
        [sg.Text('- Installation of the required python packages is manual or via `install_dependencies.py`.\n Future updates will detect missing libraries and install them automatically.\n')],
        [sg.Text('- No cancel button implementation yet. The only way to not parse is to close the app')]
    ]

    layout = [[sg.TabGroup([[  sg.Tab('Parse', parseLayout),
                               sg.Tab('How to setup Telegram', telegramLayout),
                               sg.Tab('Custom Message', messageLayout),
                               sg.Tab('About', aboutLayout)
                            ]], key='-TAB GROUP-', expand_x=True, expand_y=True),
               ]]

    window = sg.Window('Pararius.nl Alert ', layout)

    thread = None
    while True:
        event, values = window.read()

        if event == sg.WIN_CLOSED:
            if thread is not None:
                thread._stop()
            break

        if event == '-PARSE-':
            print("Parsing started")

            previousUrl = loadFileContent("previousUrl.txt")
            if values['-URL-'] == '' and previousUrl != '':
                url_input.update(loadFileContent("previousUrl.txt"))
                url = previousUrl
            else:
                url = values['-URL-']
                storeFile('previousUrl.txt', url)

            previousMsg = loadFileContent("previousMsg.txt")
            if values['-MSG-'] == '' and previousMsg != '':
                msg_input[0].update(loadFileContent("previousMsg.txt"))
                msg = previousMsg
            else:
                msg = values['-MSG-']
                storeFile('previousMsg.txt', msg)

            delay = 60
            if values['-DELAY-'] != '':
                try:
                    delay = int(values['-DELAY-'])
                except ValueError:
                    print("Invalid number entered, delay set to default value")

            if thread is None:
                thread = Thread(target=ParsePage, args=(url, msg, delay))
                thread.daemon = True
                thread.start()
            
    window.close()

if __name__ == '__main__':
    main()

